tox.ini

```diff
@@ -1,7 +1,7 @@
 [tox]
-envlist = py35, py36, py37, py38, pypy3
+envlist = py36, py37, py38, pypy3
 
 [testenv]
 commands =
-    pip install -e .[test]
+    pip install -e .[test] -U --upgrade-strategy=eager
```

Comment:
this tox configuration is wrong on many levels... prefer:

```
[tox]
envlist = 
    py38
    py37
    py36
    pypy3

isolated_build = true

[testenv]
setenv = 
  PIP_UPGRADE_STRATEGY=eager
extras = test
commands =
    python run_tests.py
```

And what's this ``run_test.py``? Couldn't you just do:

```
commands = 
   pytest tests
   sphinx-build -b doctest -d docs/_build/doctrees docs docs/_build/doctest
```
?

Reply:
run_tests.py allows for not duplicating the pytest and the sphinx-build commands across different call sites (such as this tox config and the Travis CI config).

I don't actually use tox much, but it's not clear what's preferable about this (e.g. in general I prefer explicit command line arguments over environment variables). In any case, thank you for the suggestions, I'll consider tweaking the tox config in a separate set of changes.

Thanks again for taking a look at this PR, but so far I haven't found your review very helpful. I am looking for reviewers to focus more on the type hints and in particular with the questions I called out in my comment above. Is that something you can help with?

Reply:
Why not call tox in Travis as all other projects do?

Reply:
Note that I'm author/maintainer of the tox project, so probably I'm more sensitive on this than other people would be. See a Github Actions variant of using tox https://github.com/pypa/virtualenv/blob/master/.github/workflows/check.yml#L22

Reply:
Thanks for the link!